qiskit/aqua/operators/state_functions/state_fn.py

```diff
@@ -0,0 +1,323 @@
+                other = StateFn('0' * self.num_qubits)
+            else:
+                raise ValueError(
+                    'Composition is not defined over Operators of different dimensions, {} and {}, '
+                    'respectively.'.format(self.num_qubits, other.num_qubits))
+
+        return new_self, other
+
+    def to_matrix(self, massive: bool = False) -> np.ndarray:
+        """ Return numpy vector representing StateFn evaluated on each basis state. Warn if more
+        than 16 qubits to force having to set massive=True if such a large vector is desired.
+        Must be overridden by child classes.
+
+        NOTE: THIS DOES NOT RETURN A DENSITY MATRIX, IT RETURNS A CLASSICAL MATRIX CONTAINING
+        THE QUANTUM OR CLASSICAL VECTOR REPRESENTING THE EVALUATION OF THE STATE FUNCTION ON
+        EACH BINARY BASIS STATE. DO NOT ASSUME THIS IS IS A NORMALIZED QUANTUM OR CLASSICAL
+        PROBABILITY VECTOR. If we allowed this to return a density matrix, then we would need
+        to change the definition of composition to be ~Op @ StateFn @ Op for those cases,
+        whereas by this methodology we can ensure that composition always means Op @ StateFn.
+        """
+        raise NotImplementedError
+
+    def to_density_matrix(self, massive: bool = False) -> np.ndarray:
+        """ Return matrix representing product of StateFn evaluated on pairs of basis states.
+        Must be overridden by child classes."""
+        raise NotImplementedError
+
+    def compose(self, other: OperatorBase) -> OperatorBase:
+        """ Composition (Linear algebra-style, right-to-left) is not well
+
+        defined for States in the binary function
+        model. However, it is well defined for measurements.
+        """
+        # TODO maybe allow outers later to produce density operators or projectors, but not yet.
+        if not self.is_measurement:
+            raise ValueError(
+                'Composition with a Statefunction in the first operand is not defined.')
+
+        new_self, other = self._check_zero_for_composition_and_expand(other)
+        # TODO maybe include some reduction here in the subclasses - vector and Op, op and Op, etc.
+        # pylint: disable=import-outside-toplevel
+        from qiskit.aqua.operators import CircuitOp
+
+        if self.primitive == {'0' * self.num_qubits: 1.0} and isinstance(other, CircuitOp):
+            # Returning CircuitStateFn
+            return StateFn(other.primitive, is_measurement=self.is_measurement,
+                           coeff=self.coeff * other.coeff)
+
+        from qiskit.aqua.operators import ComposedOp
+        return ComposedOp([new_self, other])
+
+    def power(self, other: int) -> OperatorBase:
+        """ Compose with Self Multiple Times, undefined for StateFns. """
+        raise ValueError('Composition power over Statefunctions or Measurements is not defined.')
+
+    # def to_density_matrix(self, massive=False):
```

Comment:
There is some to_density_method above - is this old commented code that can be scrubbed now

Reply:
Fixed